Add new Mapgen V7 floatland implementation

paramat opened the pull request:
First commit:

Floatland structure is vertically-compressed 3D noise.
Uses a lacunarity of 1.618 (the golden ratio) for high quality
noise.
Floatlands appear between user-settable Y limits, with smooth
tapering at each limit.
Simple user-settable density adjustment.
Shadow propagation is disabled in and just below floatlands, no
shadows are cast on the world surface.

/////////////////////////////////////////////////

Closes #9076 
Screenshots in following posts.

Floatlands throughout the atmosphere, between user-set y limits.
Smooth tapering to nothing at y limits.
Simple parameter for floatland density (no need to mess with noise parameters to do that).
Floatland 3D noise is compressed vertically by factor 4 to form flattened shapes.
Uses 5 octaves with lacunarity 1.618, the golden ratio, for higher quality noise and a minimum of coinciding patterns.
Shadow propagation is disabled in and just below the floatlands, no shadows cast on world surface.

Because river generation is disabled in floatland volumes, river 3D and 2D noises are not calculated, and floatland 3D noise is, so mapgen performance does not drop in floatland volumes.
Note that Mapgen V7 floatlands remain disabled by default and documented as 'highly unstable', so further tuning may occur later. I am submitting this now as MT 5.2.0 is planned to be soon.

paramat commented:
Documentation added and tested.
Due to my high confidence with this code i will merge this in a week, however i would appreciate review/testing/approval.

rubenwardy commented:
It looks very artificial to me - float lands should be islands like:

![image](https://user-images.githubusercontent.com/2122943/72751613-f8f3c280-3bb7-11ea-9977-a030b198feac.png)

>  Due to my high confidence with this code i will merge this in a week, however i would appreciate review/testing/approval/concept approval.

This is very much against the rules

LoneWolfHT commented:
I support floatlands being more natural

celeron55 commented:
Mapgen code tends to be quite simple so the code itself doesn't really matter, but the shape of floatlands should be discussed. @paramat should aim to implement the result of that discussion in this PR.

paramat commented:
Sorry, was not thinking clearly, it is not reasonable to merge this on my own approval. Edited my comment.

paramat commented:
I am deliberately avoiding the stereotyped floating island design seen in many other projects, i do not like those. The word 'should' is unreasonable, and there is more than one acceptable design approach.

Islands in a 2D ocean look a certain way, but floating islands are in a 3D volume. The stereotyped floating island design still uses the concept of intersection with a 2D surface, expressed by the rim.
So there is a lot of potential for more imaginative and less stereotyped designs.

These are designed to be consistent with the simple design of the rest of Mapgen V7, specifically the simple 3D noise that creates mountains. The mountains also look 'artificial' and unnatural.

I will wait for more core dev input before altering the design.

paramat commented:
I was already close to realising my mistake and would have changed my mind if a core dev objected, so if celeron55 was asked to intervene here (was that the case?) that was not necessary and done a little too soon. Please at least wait until i am asked and respond by refusing to back down.

Rubenwardy did object, but i am busy so did not immediately see that. The lack of a response in the 41 minutes before celeron55's post was not a sign of me refusing to back down.

paramat commented:
Hmm i have not promoted this well enough.

Here are 2 suggestions for testing that make it easier to judge the large scale structure of the floatland realm.
I also post the resulting screenshots, but keep in mind the resulting low resolution of the floatlands will make their shapes look rough and a little ugly, these screenshots are only to give an impression of large scale structure and distribution.

First 1/8th scale:
Edit the noise 'spreads' as follows:

![Screenshot from 2020-01-20 23-41-47](https://user-images.githubusercontent.com/3686677/72764953-0d4cb500-3be2-11ea-871b-58022173538a.png)

![screenshot_20200120_234954](https://user-images.githubusercontent.com/3686677/72765003-4be26f80-3be2-11ea-9a17-ee97b491506f.png)

Then 1/4 scale:

![Screenshot from 2020-01-20 23-52-27](https://user-images.githubusercontent.com/3686677/72765044-6ddbf200-3be2-11ea-84c6-f16c265ef669.png)

![screenshot_20200121_000114](https://user-images.githubusercontent.com/3686677/72765071-8ba95700-3be2-11ea-80e9-9df36d2d3946.png)

The intention with this floatland design is to not use any kind of discrete 'levels' to create the stereotyped floating island 'rims', with many floating islands sharing the same rim level.
Instead, it is designed to be 3D-homogenous, which makes sense as this is now a very tall 3D realm instead of a single 2D realm as in the previous floatland implementation.

However, i have deliberately compressed the noise vertically to create wide and flattened areas for building and exploring. So the wide areas and 'rims' are still present, but now exist branching out continuously on many levels.

It is not actually possible to code stereotyped floating islands spread homogenously in 3D space, the rim characteristic requires specific y-levels which would then apply to all floating islands.

An issue with the stereotyped floating island design is that it does not have much 'verticality', and makes travel and island hopping a mostly horizontal thing. Each level is not much different from the surface world of islands and oceans.
You can see from the screenshots above that there is a lot of potential for vertical travel by island hopping in spiral paths, or by climbing the vertical structures present in the floatlands and then moving out across the branching wide structures.
The 1/4 scale screenshot clearly shows an interesting multilevel 3D structure to the floatlands.

paramat commented:
Although i am fairly happy with the current state, i feel like spending more time tuning and experimenting, so marked as WIP again. Seems a good idea to keep tuning until we are closer to releasing 5.2.0.
If merged for 5.2.0 this feature will still be highly unstable and may be tuned further during 5.3.0 dev.

Mapgen V7 already has some floating terrain at lower altitude created by the mountain 3D noise, so it is appropriate that the floatlands have a similar design style (simple, spiky 3D noise). However, i am also making an effort to make the floatland 3D noise look as good as possible, by experimenting with noise flags and lacunarity < 2.0.

Here are some screenshots of the current state of the PR, terrain at full scale, no decorations to make the structure clear:

![screenshot_20200123_004417](https://user-images.githubusercontent.com/3686677/73026539-7d00b100-3e29-11ea-8751-d6ce700d000b.png)

![screenshot_20200123_004823](https://user-images.githubusercontent.com/3686677/73026545-80943800-3e29-11ea-9554-848e076892a9.png)

![screenshot_20200123_004114](https://user-images.githubusercontent.com/3686677/73026548-8427bf00-3e29-11ea-90a4-7eeedf28bf0c.png)

![screenshot_20200123_010348](https://user-images.githubusercontent.com/3686677/73026560-87bb4600-3e29-11ea-995e-4cce3c8e86c1.png)

paramat commented:
It was a good idea to continue experimenting. I tried normal 3D noise (no 'absvalue' flag) and prefer it, so have updated the PR. Screenshots of new parameters below.
As this is no longer 'absvalue' noise it is more consistent with 'mountain / low altitude floatland' noise.
PR is complete but still marked WIP as i might continue experimenting with parameters.

![screenshot_20200126_014611](https://user-images.githubusercontent.com/3686677/73208416-564ebd00-413e-11ea-9ea2-c71ad31b938c.png)

![screenshot_20200126_021457](https://user-images.githubusercontent.com/3686677/73208432-5ea6f800-413e-11ea-92ef-1c1187f20090.png)

![screenshot_20200126_020305](https://user-images.githubusercontent.com/3686677/73208475-6cf51400-413e-11ea-90c4-833954cacbe3.png)

paramat commented:
Tuned density, below is a 1/5th scale view:

![screenshot_20200131_190701](https://user-images.githubusercontent.com/3686677/73569545-937abe00-4462-11ea-87c3-74fde95ac864.png)

I'm happy with the tuning so removed WIP label.

paramat commented:
This should probably be merged for 5.2.0, as i wrote would happen in the news thread, such that floatlands continue from last release.

I am personally neutral about having floatlands in Mapgen V7, it is an old mapgen and i think i should have left it alone and moved on.
However, some people are using floatlands, one person needs them for their game, and other core devs insist on a replacement, so they probably should be in Mapgen V7 and in 5.2.0.
They are still documented as 'highly unstable', so this does not necessarily have to be their final form.

I am 99% confident about the code. I have spent huge amounts of time working on existing and new core mapgens. So although review may seem difficult i suggest you can somewhat trust me on the code.

paramat commented:
Moved to 5.3.0 by consensus.

Treer commented:
> Islands in a 2D ocean look a certain way, but floating islands are in a 3D volume. The stereotyped floating island design still uses the concept of intersection with a 2D surface, expressed by the rim.
So there is a lot of potential for more imaginative and less stereotyped designs.
>
>These are designed to be consistent with the simple design of the rest of Mapgen V7, specifically the simple 3D noise that creates mountains. The mountains also look 'artificial' and unnatural.

There's a charm the 2d floatlands with those biome-height shorelines have over 3d perlin noise, so... is it possible to do both? The 2d landscape in the sky at height 1250 with 3d islands floating in the space above.

paramat commented:
No it is not possible to do both, because the previous floatland generation relied on the problems that were the reasons for removing it.

paramat commented:
Bump ...
Good to see so much new dev activity happening, but please remember that along with the Android PR this was marked as a high priority PR.
As mentioned above, these floatlands are still documented as 'highly unstable', so we are free to make further changes or improvements after a merge.
Also, i am very confident about the code, this being one of my few specialities.

sfan5 commented on builtin/settingtypes.txt:
```diff
@@ -1594,12 +1594,25 @@ mgv6_np_apple_trees (Apple trees noise) noise_params_2d 0, 1, (100, 100, 100), 3
+mgv7_floatland_ymin (Floatland minimum Y) int 1024
+
+#    Upper Y limit of floatlands.
+mgv7_floatland_ymax (Floatland maximum Y) int 30000
```
I don't think the entire upper part of the world should have floatlands, 1024-4096 would be plenty without blocking the entire sky in the default configuration

paramat replied:
Good point, i am happy to reduce the volume.

paramat replied:
Done in code, will update docs later.

sfan5 commented on src/mapgen/mapgen_v7.cpp:
```diff
@@ -440,6 +471,32 @@ int MapgenV7::generateTerrain()
+	//// Floatlands
+	// Bool for optimisation of condition checks in y-loop
+	bool gen_floatlands = false;
+	float float_offset_cache[csize.Y + 2];
```
This should rather be allocated on the heap.

paramat replied:
Ok, i have no idea how, but will try to find examples elsewhere in MT code to learn from.

sfan5 replied:
like usual: `float *float_offset_cache = new float[csize.Y + 2];` here and `delete[] float_offset_cache;` to free the memory

paramat replied:
Ok thanks, will do.
(I just did a websearch, which suggested using `malloc()`, and that seemed weird.)

paramat replied:
I found examples elsewhere in code, but those always allocate in the constructor and delete in the destructor. It possibly seems better to do this here too to avoid allocating and freeing for every mapchunk?

sfan5 replied:
That would break `num_emerge_threads > 1` in a subtle way since a single `Mapgen` instance is used in multiple threads.
Allocating every time is not a problem either way, e.g. the client also makes many allocations to render a single frame.

paramat replied:
Ok thanks.
Allocation done.

sfan5 replied:
Sorry, just noticed what I said is wrong. Every thread has its own mapgen, so the caching the noise buffer as a class variable is definitely okay.
(One allocation per chunk is not too much though, so it's not <i>required</i> to cache it)

paramat replied:
Aha ok, just saw this comment, will do.

paramat replied:
Now a class variable.

sfan5 commented on src/mapgen/mapgen_v7.h:
```diff
@@ -1,7 +1,7 @@
 /*
 Minetest
-Copyright (C) 2013-2018 kwolekr, Ryan Kwolek <kwolekr@minetest.net>
```
This should stay at 2018, shouldn't it?

paramat replied:
Good question, i am not sure if copyright dates should cover the times of contributions, or cover up to the present.

sfan5 replied:
Time of contributions is also how it's been handled in other files, I think.

paramat replied:
Yes i think so too.
Searching shows kwolekr's last commits were in 2016 so used that.

sfan5 commented:
Unlike previous floatlands, these are huge.
I know that this is a design decision, but I kinda liked the more "manageable" terrain of the old floatlands.
![screenshot_20200418_214028](https://user-images.githubusercontent.com/1042418/79669733-ee989300-81bd-11ea-9666-5b97f4775c44.jpg)

> Because river generation is disabled in floatland volumes

I can see why rivers are disabled, so it would be nice if (in the future) lakes could be implemented for floatlands.

paramat commented:
Sorry, after writing that i had an idea, so WIP.

paramat commented:
> these are huge.

They are rather, i will reduce the scale.

paramat commented:
2nd commit:
'Make solid realm possible, with optional water'

Recently i saw another request in the forum for a way to have some kind of 'upper world' generated that could be used for another planet or a moon. There have been many similar requests, such as tenplus1 asking for more 'world levels'.
While i was working on this PR, i was thinking about how, if the density of these floatlands is set high in the parameters, they form a solid layer with hill-like features on the upper and lower surfaces.
I realised that there was potential for these floatlands to be optionally configured to form a solid 'upper realm'.

In the initial form of this PR, the exponent of the floatland tapering was hardcoded, so the hills on the top surface were an unsuitable shape for creating a landscape.
The tapering distance was also hardcoded, meaning that hill height was fixed.
So i added new parameters for the tapering exponent and tapering distance. With these, the landscape of the upper surface can be flexibly altered.
The resulting mapgen of a solid upper realm upper surface is a simple '3D density noise plus density gradient' method, as described early in celeron55's blog, and as used in mgv5. This type of mapgen has a 'slightly alien' look, that is suitable for the probable uses.

My second realisation was that the upper surface of a solid upper realm can safely have lakes and seas, because these will not flow downwards through gaps or flood the world surface below.
So i added a new mapgen flag 'floatwater' to enable the placing of liquid, and a parameter for the surface level of this liquid.
Of course, any game that uses a solid upper realm with liquid will have to register a special system of suitable biomes for the upper realm underground, seabed, beaches and land.

The result is that, with very little extra complexity, the floatlands can be reconfigured into a solid upper realm with optional lakes and seas.

To be clear, in this PR these new features are not used by default, because we want floatlands by default, and mgv7 had floatlands for years in the previous implementation.
Also because floatlands do not require any change to registered biomes, the normal land/mountain biomes function well, keeping compatibility with the most games and functioning with MTG.

![screenshot_20200421_191316](https://user-images.githubusercontent.com/3686677/80154456-8380ff00-85b7-11ea-85cb-5ad3a362b078.png)

![screenshot_20200420_224422](https://user-images.githubusercontent.com/3686677/80154472-8976e000-85b7-11ea-8e6b-3c6825482121.png)

paramat commented:
PR is currently buggy, attempting a fix ...

paramat commented:
Oh hang on, i might have missed something ...

paramat commented:
I have a bad feeling that, because the new flag is added after the unused 'biomerepeat', we would have to add 'biomerepeat' back into the code and in settingtypes.txt, making it appear in All Settings, .conf, map_meta.txt etc. This would be really messy because it is unused.

I am beginning to think it might be best to not add a new flag and just enable upper realm liquid according to particular values of 'floatland_ywater'.

sfan5 commented:
Adding debug output  to `readParams` reveals that it is called twice, once with the user settings and once with the map settings:
https://github.com/minetest/minetest/blob/914dbeaa0be4b5ef87506b605ef4e241cd3732dc/src/map_settings_manager.cpp#L183-L187

Since `m_map_settings` always contain the mapgen defaults, user settings will always be ignored. This has been broken since e8a8185d24897ccf964327017effae81aa1c9d40.

paramat commented:
Thanks for investigating. Big bug, good thing 5.3.0 is coming soon.

paramat commented:
Review attended too.
This is now just waiting for #9737 to be merged to be functional and testable.

sfan5 commented:
This does look interesting but I don't think it's supposed to happen (`floatland_density = 2`, `floatwater`):
![screenshot_20200425_233112](https://user-images.githubusercontent.com/1042418/80291340-04253400-874d-11ea-9ea1-f3c975a8d49b.jpg)

sfan5 commented:
> "must be configured to be a solid layer by setting 'mgv7_floatland_density' to roughly 2.0 to 3.0"

Two observations relating to this:

I have found that any density over `0.5` creates terrain where you can't advance upwards without digging through the ceiling sooner or later.
So technically that would also be enough to generate lakes at some point.

With density of `2.0` or `3.0` you get huge amounts of stone from 1024 to approximately 4096, where the surface is.
This doesn't really deserve to be called "floatlands", it's more like a second surface.

<hr />

Lastly:
```
mgv7_floatland_density = 0
mgv7_floatland_ymax = 1200
mgv7_float_taper_exp = 2.5
```
make some nice one-layer terrain with plenty holes.

paramat commented:
Now functional, but have started a 4th commit for further refinements and tuning. So WIP again.
> This does look interesting but I don't think it's supposed to happen

Just a game issue, not a mapgen issue. Liquid on a solid upper realm is a special optional bonus feature for use with games and mods that want a solid upper realm with liquid and use a suitable biome system.
MTG has just removed the problematic floatland biome code that supported a second layer of liquid for good reasons which were discussed at length in the corresponding issues and PRs.
MTG will still support a solid upper realm without liquid.
> I have found that any density over 0.5 creates terrain where you can't advance upwards without digging through the ceiling sooner or later.
So technically that would also be enough to generate lakes at some point.

I see what you mean, but ...
Density > 0.5 still has very large voids that are present below 'floatland_ymax - floatland _taper'. Liquid is only placed from 'floatland_ymax - floatland _taper' upwards, so any voids will cause vast server-killing liquid flows.
So density > 2.0 is essential to make sure the realm is completely solid below 'floatland_ymax - floatland _taper'.
Liquids can only be placed above 'floatland_ymax - floatland _taper' because if it was placed throughout the floatland volume it would inevitably be placed in a lower void that is likely to connect to the air below, causing vast flooding of the surface world.
It is also very difficult to tell from parameters whether these large voids are connected to the floatland lower limit or not.
So the only safe way to add liquid to a solid upper realm is by having a completely solid upper realm.
> This doesn't really deserve to be called "floatlands", it's more like a second surface.

Yes, i will alter the wording in documentation.

paramat commented:
No longer WIP.

4th commit:

Improve tapering behaviour:
The floatland noise offset values were scaled too large, so exponential tapering did not work as desired, now improved.

Remove floatwater flag:
Water on a solid floatland layer is a special and dangerous option that depends on carefully customised parameters and a game with a suitable biome system. So it is not a fundamental mapgen option.
I think it is best this is not a highly-visible mapgen flag, to be less tempting for beginners to enable without reading the warnings or taking the necessary precautions. Mapgen-specific flags will soon be settable in the main menu (Wuzzy's commit), where the parameter warnings are not shown.
So now the floatland water level parameter itself enables water placement according to its value.
This way also avoids further complication of mgv7 mapgen-specific flags, which are already a bit of a mess, having an unused value. I prefer to leave these flags as they are.

sfan5 commented on src/mapgen/mapgen_v7.cpp:
```diff
@@ -446,6 +486,36 @@ int MapgenV7::generateTerrain()
+	// 'Generate floatlands in this mapchunk' bool for
+	// simplification of condition checks in y-loop.
+	bool gen_floatlands = false;
+	float *float_offset_cache = new float[csize.Y + 2];
```
Please move this allocation to the constructor.

paramat replied:
Will do, just saw your comment above too.

paramat replied:
Done.

sfan5 commented:
:+1: apart from that comment

Have you thought about changing e.g. humidity with Y so that a particular (X, Z) column doesn't always have the same biome in the floatlands? I believe valleys has such a feature already.

paramat commented:
@sfan5 
5th commit makes the cache a class variable. Retested.
Perhaps this has your approval now?

sfan5 commented:
(yes)
